File: backend/schemas/prediction.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional

# Base schemas
class PredictionBase(BaseModel):
    crime_event_id: Optional[int] = None
    prediction_type: str
    prediction_value: str
    confidence_score: float = 1.0

class PredictionCreate(PredictionBase):
    pass

class PredictionResponse(PredictionBase):
    id: int
    generated_at: datetime

    model_config = {
        "from_attributes": True
    }

# Health Check schema
class HealthData(BaseModel):
    status: str
    models_loaded: Dict[str, bool]

class PredictionHealthResponse(BaseModel):
    success: bool = True
    data: HealthData

# Repeat Offender schemas
class RepeatOffenderRequest(BaseModel):
    age: float = Field(..., ge=0, le=120)
    occupation: str
    caste: str
    district: str

class RepeatOffenderData(BaseModel):
    probability: float
    risk_level: str

class RepeatOffenderResponse(BaseModel):
    success: bool = True
    data: RepeatOffenderData

# Crime Risk schemas
class CrimeRiskRequest(BaseModel):
    district: str
    crime_category: str
    historical_crime_count: int = Field(..., ge=0)

class CrimeRiskData(BaseModel):
    risk_score: float
    risk_level: str

class CrimeRiskResponse(BaseModel):
    success: bool = True
    data: CrimeRiskData

# Crime Type schemas
class CrimeTypeRequest(BaseModel):
    district: str
    month: int = Field(..., ge=1, le=12)
    hour: int = Field(..., ge=0, le=23)
    day_of_week: int = Field(..., ge=0, le=6)
    historical_crime_count: int = Field(..., ge=0)

class CrimeTypeData(BaseModel):
    crime_type: str
    confidence: float

class CrimeTypeResponse(BaseModel):
    success: bool = True
    data: CrimeTypeData

# Emerging Hotspot schemas
class HotspotRequest(BaseModel):
    district: str
    trend_metrics: float
    historical_crime_growth: float = Field(..., ge=0)

class HotspotData(BaseModel):
    hotspot_probability: float
    trend: str

class HotspotResponse(BaseModel):
    success: bool = True
    data: HotspotData

# Explainability schemas
class ExplainRequest(BaseModel):
    prediction_type: str  # repeat-offender, crime-risk, crime-type, hotspot
    features: Dict[str, Any]

class SHAPFeatureImpact(BaseModel):
    feature: str
    impact: float

class ExplainResponse(BaseModel):
    success: bool = True
    data: List[SHAPFeatureImpact]

File: frontend/types/prediction.ts
export interface PredictionResponse<T> {
  success: boolean;
  data: T;
}

export interface PredictionHealth {
  status: string;
  models_loaded: {
    repeat_offender: boolean;
    crime_type: boolean;
    crime_risk: boolean;
    hotspot: boolean;
  };
}

export interface RepeatOffenderRequest {
  age: number;
  occupation: string;
  caste: string;
  district: string;
}

export interface RepeatOffenderResult {
  probability: number;
  risk_level: "LOW" | "MEDIUM" | "HIGH";
}

export interface CrimeRiskRequest {
  district: string;
  crime_category: string;
  historical_crime_count: number;
}

export interface CrimeRiskResult {
  risk_score: number;
  risk_level: "LOW" | "MEDIUM" | "HIGH";
}

export interface CrimeTypeRequest {
  district: string;
  month: number;
  hour: number;
  day_of_week: number;
  historical_crime_count: number;
}

export interface CrimeTypeResult {
  crime_type: string;
  confidence: number;
}

export interface HotspotRequest {
  district: string;
  trend_metrics: number;
  historical_crime_growth: number;
}

export interface HotspotResult {
  hotspot_probability: number;
  trend: "RISING" | "STABLE" | "FALLING";
}

export interface SHAPImpact {
  feature: string;
  impact: number;
}

export interface ExplainRequest {
  prediction_type: "repeat-offender" | "crime-risk" | "crime-type" | "hotspot";
  features: Record<string, unknown>;
}
